h2/src/main/org/h2/table/TableFilter.java

```diff
@@ -709,18 +714,32 @@ public void setNestedJoin(TableFilter filter) {
      * @param on the condition
      */
     public void mapAndAddFilter(Expression on) {
+        this.onCondition = on;
         on.mapColumns(this, 0, Expression.MAP_INITIAL);
         addFilterCondition(on, true);
-        on.createIndexConditions(session, this);
         if (nestedJoin != null) {
             on.mapColumns(nestedJoin, 0, Expression.MAP_INITIAL);
-            on.createIndexConditions(session, nestedJoin);
         }
         if (join != null) {
             join.mapAndAddFilter(on);
         }
     }
 
+    public void createIndexConditions() {
+        if (onCondition != null) {
+            onCondition.createIndexConditions(session, this);
+            if (nestedJoin != null) {
+                onCondition.createIndexConditions(session, nestedJoin);
+            }
+        }
+        if (join != null) {
+            join.createIndexConditions();
+        }
```

Comment:
this block might need to got at the end of the method, which might make the new plans more like the old plans, but I suspect you are largely right, and that the re-ordering is of no consequence

Reply:
In the first commit (with separate `on` condition) there are no changes in plans in our test suit.

Plans were changed only in the second commit, where I reused `joinCodition`.